test/Salsa.jl

```diff
@@ -159,6 +159,23 @@ end
     @test x(rt, 1, "hi") == "ho"
 end
 
+@testset "macro usage corner cases" begin
+    @testset "where clauses" begin
+        # A simple derived function with a where clause
+        Salsa.@derived where_func_x(db, x::T) where T = sizeof(T)
+        @test where_func_x(Runtime(), 0) == 8
+        @test where_func_x(Runtime(), Int8(0)) == 1
+
+        # Derived function with typed argument
+        Salsa.@derived where_func_T(db, ::Type{T}) where T = sizeof(T)
+        # TODO: This is broken because typeof(Int) is DataType, but DataType is not <: Type
+        @test_broken where_func_T(Runtime(), Int) == 8
```

Comment:
This test still doesn't work because of the garbage way we wrote `get_user_function(::Runtime, ::DependencyKey)`.

Basically, we need to be able to figure out which function to call just from the DependencyKey alone. Currently we do that by adding a method to `get_user_function()` for each new derived function that's defined, which specifies which user function to call for a given DependencyKey. Really, we only need to do this mapping once, because no matter what the types are, every `DependencyKey{typeof(foo)}` is always going to map to `%%__user_foo` or whatever. _BUT_ we can't dispatch only on the function type, because then we'd be defining redundant methods if you have multiple methods for a derived function, and then julia will complain about you replacing a method definition. So we're defining it with the full DependencyKey type, just to keep them separate, even though it's redundant. BUT THEN, it's hard to get the DependencyKey type right when the arguments are types, like this, because `typeof((Int,))` ends up returning `Tuple{DataType,}`, but the user wrote `Type{T} where T` in the function definition -_-

Reply:
Oh, actually, NVM -- writing this explanation was enough for me to see the fix! :) I fixed it by adding a `hasmethod()` to skip adding the new method if one already exists! :))